sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -63,15 +63,26 @@ case class Concat(children: Seq[Expression]) extends Expression with ImplicitCas
 
   override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val evals = children.map(_.genCode(ctx))
-    val inputs = evals.map { eval =>
-      s"${eval.isNull} ? null : ${eval.value}"
-    }.mkString(", ")
-    ev.copy(evals.map(_.code).mkString("\n") + s"""
-      boolean ${ev.isNull} = false;
-      UTF8String ${ev.value} = UTF8String.concat($inputs);
-      if (${ev.value} == null) {
-        ${ev.isNull} = true;
+    val argNums = evals.length
+    val args = ctx.freshName("argLen")
+    ctx.addMutableState("UTF8String[]", args, "")
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      if (eval.isNull != "true") {
+        s"""
+          ${eval.code}
+          $args[$index] = ${eval.value};
```

Comment:
If `eval.isNull` is evaluated to `null` at runtime, `eval.value` is useless. We should assign `null` in that case.

Reply:
good catch